Document for UPD-12, UPD-30, UPD-33, and improve docs for deps upgrade checks (#870)
digital-asset/docs.daml.com: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst

Package authors may upgrade the dependencies of a package as well as the
package itself. A new version of a package may add new dependencies, and
must have all the (non-:ref:`utility-package <upgrades-utility-package>`)

> dependencies of the old version. When using these dependencies in ways that are
> checked for upgrades, each of these existing dependencies must either be
> unchanged from the old dar, or an upgrade of its previous version.

Reviewer: ```suggestion
dependencies of the old version. If these dependencies are used in ways that are
checked for upgrades, each existing dependency must be either
unchanged from the old DAR or an upgrade of its previous version.
```

> For example, given a dependencies folder, containing v1, v2, and v3
> of a dependency package ``dep``:

Reviewer: ```suggestion
For example, given a dependencies folder containing v1, v2, and v3
of a dependency package ``dep``:
```

  - daml-prim
  - daml-stdlib
  - daml3-script
  data-dependencies:
  - dependencies/dep-3.0.0.dar # Can upgrade dep-2.0.0 to dep-3.0.0, or leave it unchanged
  ...

> Downgrading a dependency when using that dependency's datatypes is not
> permitted. For example, ``main`` may not downgrade ``dep`` to version ``1.0.0``.
> The following ``daml.yaml`` would be invalid:

Reviewer: ```suggestion
You cannot downgrade a dependency when using that dependency's datatypes. For example, ``main`` may not downgrade ``dep`` to version ``1.0.0``.
The following ``daml.yaml`` would be invalid:
```

  - daml-prim
  - daml-stdlib
  - daml3-script
  data-dependencies:
  - dependencies/dep-1.0.0.dar # Cannot downgrade dep-2.0.0 to dep-1.0.0
  ...

> If you try to build this package, the typechecker will error on a package ID
> mismatch for the Dep:AdditionalData field, because the Dep:AdditionalData
> reference in this case has changed to a package that is not a legitimate upgrade
> of the original.

Reviewer: ```suggestion
If you try to build this package, the typechecker returns an error on a package ID
mismatch for the Dep:AdditionalData field, because the Dep:AdditionalData
reference in this case has changed to a package that is not a legitimate upgrade
of the original.
```

version 2 package also causes issues, especially if the interface has
choices.

> This means that template definitions that exist in the same package as
> interfaces and exception definitions will not be upgradeable. To avoid this
> issue, move interface and exception definitions out into a separate package in
> your model, such that subsequent versions of your package with templates all
> depend on the same version of the package with interfaces/exceptions.

Reviewer: ```suggestion
This means that template definitions that exist in the same package as
interfaces and exception definitions are not upgradeable. To avoid this
issue, move interface and exception definitions into a separate package such that subsequent versions of your template package all depend on the same version of the package with interfaces/exceptions.
```

For example, a single package ``main`` defined as follows would not be able to
upgrade, leaving the template ``T`` non-upgradeable.

.. code:: daml

  module Main where

  interface I where
    ...

  template T with
    ...

.. code:: yaml

  sdk-version: 2.10.0
  name: main
  version: 1.0.0
  source: Main.daml
  dependencies:
  - daml-prim
  - daml-stdlib
  build-options:

>   - --target=1.dev

Reviewer 2: I think I left this on the commit before, so leaving it here too.
This should be 1.17

> The SCU type checker will emit an error and refuse to compile this module:

Reviewer: ```suggestion
The SCU type checker emits an error and refuses to compile this module:
```

state can then only be used by explicitly choosing the older version of
the contract in your transaction.

Breaking Changes via Explicit Package Version
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> In the event that you would like to make a breaking change to your package that
> is not upgrade compatible, you can change the name of your package to indicate a
> breaking version bump. In order to enable this, we recommend that your package
> name contains a version marker for when a breaking change does happen.

Reviewer: ```suggestion
To make a breaking change to your package that
is not upgrade compatible, you can change the name of your package to indicate a
breaking version bump. To enable this, we recommend that your package
name contains a version marker for when a breaking change occurs.
```

For example, for your first iteration of a package, you would name it
``main-v1``, starting with package version ``1.0.0``. In this case, the ``v1``
is part of the *package name*, not the package version. You could publish
upgrade-compatible versions by changing the ``version:`` field from ``1.0.0`` to
``2.0.0`` to ``3.0.0``. These versions would all be upgrade-compatible with
one another:

.. code:: text

  main-v1-1.0.0
  main-v1-2.0.0
  main-v1-3.0.0

Note how the ``v1`` in all three packages remains stable - this means the
package name has not changed, and ensures that these three packages and their
datatypes are considered by the runtime and the type checker to be upgradeable.

When you want to make a breaking change, you would publish a new version of the
package with package name ``main-v2``. Because this package would have a
different package name from those with ``main-v1``, it would not be typechecked
against those packages and its datatypes would not automatically be converted.

> You would need to manually convert values from ``main-v1`` packages to be

Reviewer 2: Perhaps mention using a migration tool like the upgrade tool

> When you want to make a breaking change, you would publish a new version of the
> package with package name ``main-v2``. Because this package would have a
> different package name from those with ``main-v1``, it would not be typechecked
> against those packages and its datatypes would not automatically be converted.
> You would need to manually convert values from ``main-v1`` packages to be

Reviewer: ```suggestion
To make a breaking change, you would publish a new package version with the package name ``main-v2``. Because this package would have a
different package name from those with ``main-v1``, it would not be typechecked
against those packages and its datatypes would not automatically be converted.
You would need to manually convert values from ``main-v1`` packages to be
```

Testing
=======

Standalone Upgradeability Checks
--------------------------------

> We recommend that as a further check for the validity of your upgraded packages,
> you perform a standalone check that all of the DARs typecheck against one
> another correctly via the ``upgrade-check`` tool.

Reviewer: ```suggestion
We recommend using the ``upgrade-check`` tool to perform a standalone check that all of the DARs typecheck against one another correctly as further validation of your upgraded packages.
```

> This tool takes a list of DARs and runs Canton's participant-side upgrade
> typechecking, without spinning up an instance of Canton. You should pass the
> tool the list of DARs constituting your previous model and the list of DARs for
> your new model.

Reviewer: ```suggestion
This tool takes a list of DARs and runs Canton's participant-side upgrade
typechecking without spinning up an instance of Canton. You should pass the
tool the list of DARs constituting your previous model and the list of DARs for
your new model.
```

For example, assume you have a helper package ``helper`` that does not change,
and two packages ``main`` and ``dep``.

.. code:: text

  main-1.0.0.dar
  dep-1.0.0.dar
  helper-1.0.0.dar

> After upgrading your model, you would a new DAR ``main-2.0.0.dar`` for ``main``
> and a new DAR ``dep-2.0.0.dar`` for ``dep``. We would then recommend running the
> upgrade-check tool as follows:

Reviewer: ```suggestion
After upgrading your model, you would publish a new DAR ``main-2.0.0.dar`` for ``main``
and a new DAR ``dep-2.0.0.dar`` for ``dep``. We would then recommend running the
upgrade-check tool as follows:
```

.. code:: bash

  > daml upgrade-check --participant helper-1.0.0.dar dep-1.0.0.dar main-1.0.0.dar dep-2.0.0.dar main-2.0.0.dar
  ...

> This will run the same upload validation over these DARs that would be run in
> the event of an upload to the ledger, and will print out the same messages and
> errors. Because it will not require a ledger to be spun up, the command runs
> much more quickly.

Reviewer: ```suggestion
This runs the same upload validation over these DARs that would be run in
the event of an upload to the ledger, and prints out the same messages and
errors. Because it does not require a ledger to be spun up, the command runs
much more quickly.
```

We can also check that all of the DARs pass compiler-side checks, but this is
much less likely to indicate an issue because the DARs are typechecked during
compilation anyways.

.. code:: bash

  > daml upgrade-check --compiler helper-1.0.0.dar dep-1.0.0.dar main-1.0.0.dar dep-2.0.0.dar main-2.0.0.dar
  ...

Dry Run Uploading to a Test Environment
---------------------------------------

If you have a test environment with DARs that are not available to you, you may
not be able to supply a complete list of DARs for your previous model to the
standalone ``upgrade-check`` tool.

In this case, we recommend that as a further check for the validity of your
upgraded package, you perform a dry-run upload of your package to a testing
environment, using the ``--dry-run`` flag of the ``daml ledger upload-dar``

> command. This also runs the upgrade type-checking, but does not persist your

Reviewer: ```suggestion
command. This also runs the upgrade typechecking, but does not persist your
```

This tool takes a list of DARs and runs Canton's participant-side upgrade
typechecking, without spinning up an instance of Canton. You should pass the
tool the list of DARs constituting your previous model and the list of DARs for
your new model.

For example, assume you have a helper package ``helper`` that does not change,
and two packages ``main`` and ``dep``.

.. code:: text

  main-1.0.0.dar
  dep-1.0.0.dar
  helper-1.0.0.dar

After upgrading your model, you would a new DAR ``main-2.0.0.dar`` for ``main``
and a new DAR ``dep-2.0.0.dar`` for ``dep``. We would then recommend running the
upgrade-check tool as follows:

.. code:: bash

  > daml upgrade-check --participant helper-1.0.0.dar dep-1.0.0.dar main-1.0.0.dar dep-2.0.0.dar main-2.0.0.dar
  ...

This will run the same upload validation over these DARs that would be run in
the event of an upload to the ledger, and will print out the same messages and
errors. Because it will not require a ledger to be spun up, the command runs
much more quickly.

We can also check that all of the DARs pass compiler-side checks, but this is
much less likely to indicate an issue because the DARs are typechecked during

> compilation anyways.

Author: I opted to drop the word `anyways`.
